### ISSUE TITLE
ErrorInterceptorHandler(handler) doesn't work after unlocking dio

### ISSUE THREAD

**Issue by Okiring** [flutterchina/dio#1272]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | android                      |      |
| Platform Version               |2.0.5                              |      |
| Dio Version                    | 4.0.0                              |      |
| Android Studio             |4.3.1            |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |      |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

am using an interceptor and discovered something very odd during the usage. Am using the interceptor for refreshing the access token incase it expires. I only retry the request once and if that fails, am using the handler.next to complete the error..
![Screen Shot 2021-09-08 at 5 41 12 PM](https://user-images.githubusercontent.com/27002476/132530968-879d38a2-5bdc-40a9-b284-63450238e954.png)

when an error occurs without need to refresh the token ( lock , refresh the token and unlock the dio instance, then remake the previous request as shown in the pic), the handler.next function works fine and is able to pass back the error to my try-catch block...HOWEVER, when the access-token is refresh and a new request is made with a valid token, incase the new request fails with an error e.g 4xx,5xx or any error, the handler.next function fires but the error is never caught by the try-catch block
![Screen Shot 2021-09-08 at 5 22 19 PM](https://user-images.githubusercontent.com/27002476/132529617-6178a547-28dc-4ea1-9207-a8f48c5f0831.png)
![Screen Shot 2021-09-08 at 5 23 24 PM](https://user-images.githubusercontent.com/27002476/132529657-25d5147a-8941-494a-91f8-bfaf19417ff3.png)
